Hapi Riot Mark 1

jrans opened the pull request:
Expand on first simple idea of using `Riot.render` to create a view engine by
- **MOST IMPORT** by mounting tag instances on client side with correct context to replace initially rendered html when js available #13 
- adding layout features #16
- improving caching #19

Instructions given and examples updated while todo example left as WIP.

@nelsonic Breaking changes so will have new a big bump on the version when merged in.

Next steps: #24 and to just use it and find more issues!

jrans committed:
:bulb: Adding client side javascript ideation #13

jackcarlisle committed:
updates the readme with instructions on including javascript and css compiled #18

jrans committed:
:heavy_plus_sign: Layout feature
+ Create layout feature to have foundation html to be inject with riot content.
+ Cache this process
+ Update existing cacheing to make slightly more efficient
+ Update examples to use layout feature
Related: #16, #19, #20

nelsonic commented on .gitignore:
```diff
@@ -41,3 +41,5 @@ db
 #goodparts symlink
 .eslintrc.js
+
+compiled.js
```
for "convention" (see: http://browserify.org/ ) would it be an _idea_ to _name_ the compiled file `bundle.js` ? 💡

jrans replied:
will change and advise to use bundle.js

nelsonic commented on example/new/server.js:
```diff
@@ -0,0 +1,46 @@
+var Path = require('path');
+
+var server = new Hapi.Server();
+var port = process.env.PORT || 8000;
```
do you get _coverage_ on this line?

jrans replied:
now not covering all code in examples folder as more work than needed

nelsonic commented:
@jrans this is looking _promising_. please let me know if you need any input. 👍

jackcarlisle committed:
adds compiled file path to the compileOptions #18

jackcarlisle committed:
adds compiledFileRoute to the compileOptions #18

jackcarlisle committed:
resolving merge conflicts #18

jackcarlisle committed:
fixes markdown #18

jackcarlisle committed:
changes the layout filenames from .tag to .html #18

jackcarlisle committed:
adds correct layout.html file to todo example #18

jackcarlisle committed:
fixes compiledFileRoute variable #18

jrans committed:
:snowflake: .istanbul.yml added to exclude example folder

jackcarlisle committed:
adds work in progress to the todo example. Needs refactor. #18

jackcarlisle committed:
updates the readme #18

jrans committed:
:white_check_mark: Get coverage back up for all work done

jrans committed:
Merge branch 'layout' of https://github.com/dwyl/hapi-riot into layout

jackcarlisle committed:
Merge pull request #21 from dwyl/layout

Layout feature

jackcarlisle committed:
removes new example #18

jackcarlisle committed:
resolves branch conflicts and updates examples #18

jackcarlisle committed:
adds compile scripts to the package.json #18

jrans committed:
Merge pull request #25 from dwyl/layoutUpdate

Layout update

jrans committed:
:snowflake: Updates to package.json and basic example

jrans committed:
:memo: Document new features fully

jrans committed:
:memo: :art: Update examples
+ use bundle.js naming convention for best practises.
+ Instruct them to use npm start to see basic example

jrans commented:
@nelsonic Assigning over so you can have a look at code and play with locally but CI causing problems... Not sure why as seems like not starting as getting this error: 
<img width="884" alt="screen shot 2016-10-18 at 16 42 57" src="https://cloud.githubusercontent.com/assets/12045918/19485008/8fe3da24-9551-11e6-8e63-aabc0ba6168f.png">

nelsonic commented on README.md:
```diff
@@ -66,102 +66,190 @@ Server-side rendering of Riot Components using the [_Vision_](https://github.com
+    <link rel="stylesheet" type="text/css" href="stylesheet.css"/>
+  <head>
+  <body>
+    <<<RIOT>>>
```
I know we _discussed_ this _briefly_ but what about using a `<yield to="content">` to keep it compatible?
see: https://martinmuzatko.github.io/riot-cheatsheet/#templating-yield

jackcarlisle replied:
@nelsonic do you mean that we should replace our `layout.html` file with a riot `.tag` file that yields to our content?

nelsonic commented on lib/index.js:
```diff
@@ -44,29 +59,43 @@ function unCache (cache) {
     }
 
-    return output;
+    return '<!DOCTYPE html>' + content;
```
previously we allowed the `'<!DOCTYPE html>'` to be _overridden_? but my reading of MDN: https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/HTML5/Introduction_to_HTML5#Declaring_that_the_document_contains_HTML5_mark-up_with_the_HTML5_doctype is that this is not _necessary_ because HTML5 always has the simplified Doctype.

nelsonic commented on README.md:
```diff
@@ -66,102 +66,190 @@ Server-side rendering of Riot Components using the [_Vision_](https://github.com
 
 
-## Lessons Learned & "Gotchas"
+#### `compiledFileRoute` **REQUIRED**
```
@jrans this should have a "_sensible default_". remember we are creating this for "us" first of all. if someone else wants to use it and they want to _set_ a _different_  `compiledFileRoute` that's entirely up to them. 😉

jackcarlisle committed:
adds default route to compiledFileRoute #18

jackcarlisle committed:
removes DOCTYPE html from output #18

jackcarlisle commented:
@nelsonic made some changes from your comments and added a comment to one of them -> https://github.com/dwyl/hapi-riot/pull/17#pullrequestreview-5043291

nelsonic commented:
@jrans & @jackcarlisle I've run this _locally_ and think you've done a _sterling_ job of improving this project! 😍 
Not a _huge_ fan of the `<<<RIOT>>>` placeholder because it's non-semantic markup, _however_ it's an _internal_ convention that will never be sent to a browser so _for now_ it's fine. 👍 
Hope to see the Todo example https://github.com/dwyl/hapi-riot/issues/26 working _soon_...  😉
